plugins: Introduce phased plugin

oshoval opened the pull request:
Phased plugin allows running jobs that will be considered as "phase 2"
only once both lgtm / approve are presented.

We are running 3 k8s versions for each sig on kubevirt, for each PR.
With this plugin we can declare the non latest one as "phase 2"
and run it only when the PR is entering the merge queue,
This will save CI resources.

A job will be declared as "phase 2",  if it has `always_run=false`, `optional=false`,
without additional conditions such as `run_if_changed`, `skip_if_only_changed`.

Since the jobs are `always_run=false` it means that a potential race condition might happen
between running the job and merging.
For this we use `require_manually_triggered_jobs` which will block the PR entering 
the merge queue / being merge by tide before all "phase 2" jobs finish.
It will also cause tide to auto run phase 2 when rerunning batches.

Notes: 
* Doesn't support branches, since for branches we consider everything as stable phase 1.
A follow-up [PR](https://github.com/kubevirt/project-infra/pull/3126) will update the job forking to update phase 2 to be considered phase 1 on new branches.
* Works only for `kubevirt/kubevirt`.
* `pull-kubevirt-e2e-k8s-1.27-ipv6-sig-network` was chosen as phase 2 candid.

oshoval commented:
/test pull-project-infra-prow-deploy-test  bla

(just checking something)

oshoval commented on github/ci/prow-deploy/files/jobs/kubevirt/project-infra/project-infra-postsubmits.yaml:
```diff
@@ -26,6 +26,32 @@ postsubmits:
+              requests:
+                memory: "8Gi"
+              limits:
+                memory: "8Gi"
```
need to run privileged like #3119 maybe

oshoval replied:
@brianmcarey wdyt please?

brianmcarey reviewed (commented):
Thanks @oshoval for this - it could be very useful. I just have a couple of questions inline.

oshoval commented on github/ci/prow-deploy/files/jobs/kubevirt/project-infra/project-infra-postsubmits.yaml:
```diff
@@ -26,6 +26,32 @@ postsubmits:
+        preset-kubevirtci-quay-credential: "true"
+      spec:
+        containers:
+          - image: quay.io/kubevirtci/bootstrap-legacy:v20221205-e30b34a
```
hmm we are using bootstrap-legacy, you mean we should use bootstrap ? 
i just copied from other postsubmit

brianmcarey replied:
I would prefer to use the more update to date bootstrap image - it should work if you are just building and publishing an image?

oshoval commented on github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:
```diff
@@ -797,7 +798,7 @@ presubmits:
         securityContext:
           privileged: true
-  - always_run: true
+  - always_run: false
```
This one is a candid to be phase 2
this why i changed this one
once plugin is tested, we can change all non latest k8s version / ipv6 etc to be phase 2
unless we want ipv6 to be considered phase 1, but imo not needed
whoever works on specific ipv6 feature, which is rare, will be able to /test manually during development
chosen it as it is one that doesnt change much, and we have the dual stack running as phase 1

brianmcarey replied:
ok thanks @oshoval

oshoval replied:
Is it fine that we require to add `# phase2` comment on those  ?
this way it will be more explicit they are phase2 , and will allow us to change them to true when
creating a new release

otherwise it means we need to read the config, and filter the jobs that need to be updated
as this plugin does real time, when we create a new release

oshoval commented on external-plugins/phased/plugin/handler/handler.go:
```diff
@@ -0,0 +1,370 @@
+	log.Infof("Handling updated pull request: %s [%d]", event.Repo.FullName, event.PullRequest.Number)
+
+	if !h.shouldActOnPREvent(event) {
+		log.Infoln("Skipping event. Not of our interest.")
```
just copy pasted from rehearse, 
i can change the text, or maybe lets even remove this line ? it will just spam log
and not interesting

brianmcarey replied:
Yes better to remove it.

oshoval commented on external-plugins/phased/plugin/handler/handler.go:
```diff
@@ -0,0 +1,370 @@
+	}
+
+	// Until we update k8s.io/test-infra which allows to read require_manually_triggered_jobs policy
+	if !(org == "kubevirt" && repo == "kubevirt") && !(org == "foo" && repo == "bar") {
```
we want the plugin to do it only on org/repo/branches that have `require_manually_triggered_jobs` enabled
so lets start with hardcoded just for kubevirt (and foo/bar which is for unit tests)
once we update `k8s.io/test-infra` in project-infra we can do code that will detect if the PR org/repo has this feature enabled

brianmcarey replied:
Can we not create PR events for the kubevirt/kubevirt repo in the tests rather than including foo/bar here?

oshoval replied:
kubevirt/kubevirt is for the real use
foo/bar is for the unit tests
you mean to create on kubevirt/kubevirt on the unit tests as well and then to remove foo/bar here ? if so we can sure

brianmcarey replied:
yes thats what I mean - i don't think we need foo/bar here.

oshoval commented on github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml:
```diff
@@ -1798,6 +1799,7 @@ presubmits:
       nodeSelector:
         type: bare-metal-external
   - always_run: false
+    optional: true
```
@brianmcarey 
please note i changed `pull-kubevirt-e2e-k8s-1.27-sig-compute-realtime`
so it wont be considered phase 2

oshoval commented:
Rebased only

oshoval commented on external-plugins/phased/plugin/BUILD.bazel:
```diff
@@ -0,0 +1,72 @@
+    ],
+)
+
+load("@io_bazel_rules_docker//go:image.bzl", "go_image")
```
rules_docker is used all over project-infra (even on the new plugin bot-review)
it is better to change them all in one PR imo ?

brianmcarey replied:
Yes sorry - just a note. We will have to do this in kubevirt/kubevirt too. Its ok for now.

oshoval commented:
Remove branch support

on a follow PR we should update new branch forking to change phase 2 to be phase 1 for branches

oshoval commented:
pull-project-infra-prow-deploy-test  fails
i dont think its my PR ? it passed before small changes / rebase
or is it ?

oshoval commented:
/retest

oshoval commented:
worked now, there was a flake twice today

oshoval commented on external-plugins/phased/plugin/main.go:
```diff
@@ -0,0 +1,201 @@
+		"jobs-config-base",
+		"",
+		"Base path to a directory with Prow job configs (required).")
+	fs.StringVar(&o.jobsNs,
+		"jobs-namespace",
+		"",
+		"The namespace in which Prow jobs should be created.")
```
can be removed

oshoval commented:
first push - cosmetics: remove unused code
2nd push - rebase

oshoval commented on github/ci/prow-deploy/kustom/overlays/ibmcloud-production/kustomization.yaml:
```diff
@@ -29,6 +29,12 @@ patches:
       kind: Deployment
       name: rehearse
     path: patches/JsonRFC6902/prow-rehearse-deployment.yaml
+  - target:
```
maybe worth to delete whole folder on a follow PR ?

brianmcarey reviewed (commented):
/lgtm

thanks for this @oshoval  - lets try to get this deployed soon so we can try it out.

oshoval commented:
Thanks Brian, i will update the PR with the new image now

oshoval commented:
rebased
updated image to reflect current code

oshoval commented:
looks like a flake

/test pull-project-infra-prow-deploy-test

oshoval commented on github/ci/prow-deploy/kustom/base/manifests/local/prow-phased-deployment.yaml:
```diff
@@ -0,0 +1,57 @@
+          imagePullPolicy: IfNotPresent
+          args:
+            - --github-token-path=/etc/github/oauth
+            - --github-endpoint=http://ghproxy
```
seems we can remove this line because it is duplicate and the next line will be the actual value
(copy paste from rehearse plugin) but not critical, just cosmetics

oshoval commented on github/ci/prow-deploy/kustom/overlays/kubevirt-prow-control-plane/patches/JsonRFC6902/prow-phased-deployment.yaml:
this file (and the ref to it) can be removed, because we dont create jobs

oshoval commented:
removed uneeded patch, and double arg from manifests
rebased

oshoval commented:
add missing error check, removed unused var
rebased
update image for those changes

oshoval commented:
/test pull-project-infra-prow-deploy-test

brianmcarey commented:
> /test pull-project-infra-prow-deploy-test

The prow-deploy-test failure is unrelated to this PR - see https://github.com/kubevirt/project-infra/pull/3163

dhiller commented on external-plugins/phased/plugin/phased_test.go:
```diff
@@ -0,0 +1,201 @@
+			}
+		})
+
+		Context("lgtm label is added", func() {
```
This is the happy path, how about testing the opposite also?

dhiller commented on external-plugins/phased/plugin/handler/handler.go:
```diff
@@ -0,0 +1,367 @@
+		return false, err
+	}
+
+	return (eventLabel == labels.LGTM && github.HasLabel(labels.Approved, l)) ||
+		(eventLabel == labels.Approved && github.HasLabel(labels.LGTM, l)), nil
```
nit

```suggestion
	return github.HasLabel(labels.LGTM, l) ==  && github.HasLabel(labels.Approved, l)
```

oshoval replied:
seems like typo ?

dhiller commented on external-plugins/phased/plugin/handler/handler.go:
```diff
@@ -0,0 +1,367 @@
+		return
+	}
+
+	presumbits, err := h.loadPresubmits(*pr)
```
It would be more efficient to load the config on plugin startup and restart plugin when configmap changes. But that can be done in a follow up to avoid premature optimization.

oshoval replied:
we cant restart the plugin, it is prone for races, when the plugin is down, events can be lost 
(assuming the plugin doesnt get all the history events)

anyhow we can do a follow up optimization that checks if the file was changed indeed (according git log etc),
and load only in this case, the files doesn't change a lot

dhiller commented on external-plugins/phased/plugin/handler/handler.go:
```diff
@@ -0,0 +1,367 @@
+		return
+	}
+
+	presumbits, err := h.loadPresubmits(*pr)
```
```suggestion
	presubmits, err := h.loadPresubmits(*pr)
```

dhiller reviewed (approved):
@oshoval thank you! Looks very good from a first pass, couple of nits and one spelling mistake inline.

/approve

/hold can be removed if you think then nits aren't worthwhile.

dhiller reviewed (commented):
One thing I missed: you need to enable this plugin in the external_plugins config like here: https://github.com/kubevirt/project-infra/blob/0653b9da37d8ef1f0d71838d52a25facc090aff2/github/ci/prow-deploy/kustom/base/configs/current/plugins/plugins.yaml#L522

oshoval commented:
> One thing I missed: you need to enable this plugin in the external_plugins config like here:
> 
> https://github.com/kubevirt/project-infra/blob/0653b9da37d8ef1f0d71838d52a25facc090aff2/github/ci/prow-deploy/kustom/base/configs/current/plugins/plugins.yaml#L522

It is here isn't it ?
https://github.com/kubevirt/project-infra/pull/3103/files#diff-3a088ffbe61dd30f930cd1855ab72a652fcea4a33c4b047eeb6df38f7a79ad51R531

oshoval commented on external-plugins/phased/plugin/main.go:
```diff
@@ -0,0 +1,172 @@
+func setupLogger() *logrus.Logger {
+	l := logrus.New()
+	l.SetFormatter(&logrus.TextFormatter{FullTimestamp: true, TimestampFormat: time.RFC1123Z})
+	l.SetLevel(logrus.TraceLevel)
```
dont think we need atm, even rehearse doesnt have it, and it is copied from there
upon needs we can

oshoval commented:
addressed the marked comments (commented on the rest)
rebased
updated image

thanks

/hold cancel
(we can do the rest in a follow up as mentioned)

dhiller commented on github/ci/prow-deploy/kustom/base/configs/current/plugins/plugins.yaml:
```diff
@@ -528,6 +528,10 @@ external_plugins:
     endpoint: http://botreview:9900
     events:
       - pull_request
+  - name: phased
```
This configuration tells `hook` that it should send pull request events for the `kubevirt/project-infra` repository to the `phased` external plugin.

You can do either of these:
* move the configuration up into the `kubevirt` org section [here](https://github.com/kubevirt/project-infra/pull/3103/files#diff-3a088ffbe61dd30f930cd1855ab72a652fcea4a33c4b047eeb6df38f7a79ad51R500)
* create a new kubevirt/kubevirt section and move the config there.

dhiller replied:
```suggestion
  kubevirt/kubevirt
  - name: phased
```

This would enable the event handling explicitly for kubevirt/kubevirt repo only.

oshoval replied:
moved it to the current kubevirt/kubevirt
is it ok ?

https://github.com/kubevirt/project-infra/compare/0883ae5c8a814d25c1fd7c5b6500f6a1111e600e..c8e7d7a5c351e912b20d3de94018fa6bccee66ec

dhiller commented on external-plugins/phased/plugin/handler/handler.go:
```diff
@@ -0,0 +1,367 @@
+	}
+
+	orgRepo := org + "/" + repo
+	var presumbits []config.Presubmit
+	for index, jobs := range pc.PresubmitsStatic {
+		if index != orgRepo {
+			continue
+		}
+		for _, job := range jobs {
+			presumbits = append(presumbits, job)
+		}
+	}
+
+	return presumbits, nil
```
```suggestion
	var presubmits []config.Presubmit
	for index, jobs := range pc.PresubmitsStatic {
		if index != orgRepo {
			continue
		}
		for _, job := range jobs {
			presubmits = append(presubmits, job)
		}
	}

	return presubmits, nil
```

dhiller reviewed (changes requested):
Sorry, I overlooked the external plugin misconfiguration in first pass. That needs to get fixed.

Other than than just a spelling mistake.

oshoval commented:
Thanks

fixed comments
rebased
updated image

oshoval commented:
created kubevirt/kubevirt for the plugin
